### ISSUE TITLE
Allow arbitrary axis tick labels (especially for time-based data)

### ISSUE THREAD

**Issue by philer** [goat1000/SVGGraph#6]
I'm using a simple line graph to visualize the accumulation of certain items over time. I have one integer per day for a time range of over two years, so that's around 800 entries. They key for each day is a standard DateTime string. In the screenshot below I used `'grid_division_h' => 100`.

As you can see the full datetime strings are used as tick labels, causing some ugly overlapping. Also they are generally unreadable.

So my first approach was to map the datetime strings to a shorter format like `Jan/16` and was hoping that SVGGraph would just magically understand that duplicate keys mean each key should be used as a grid separator. Obviously that doesn't work it just complains about "Repeated keys in data". (I'm using `structured_data`, which is why duplicate keys are technically possible.)

![stats.png](https://i.imgur.com/b4PDZ3e.png)

**Comment by goat1000** [comment]
A date/time based axis is on the top of my to-do list, but until then there are a couple of options available:

1. Use the structured data `axis_text` option to set the text labels in the data;
2. Use the `axis_text_callback_x` option to set a callback function for reformatting the axis text.

Option 2 is more efficient, the function is only called for values that will appear on the axis.

**Comment by philer** [comment]
Thanks, I overlooked the `axis_text` option in the docs. The `axis_text_callback_x` doesn't appear to be documented yet and I'm not quite sure how to use it (apparently it's called with just an index, but the return value is being ignored).

**Comment by goat1000** [comment]
`axis_text_callback_x` is described in the "Grid-based graph options" table, but I might not have documented it very well:

> Name of callback function for formatting X-axis text. The function will be called with one argument: callback($value). The function should return a text string to display on the axis.

 If you use both `axis_text` and `axis_text_callback_x`, the `axis_text` value will take precedence (structured data fields always take precedence over global options).

**Comment by philer** [comment]
Can't find that table. Though I'm guessing that means I can't actually use a Closure here? I tried this:
```php
'axis_text_callback_x' => function($id) {
    return "foo";
}
```
and it's getting ignored (without having specified any `axis_text`). It just uses the key instead.

**Comment by goat1000** [comment]
There appears to be a bug there - the callback option will support any callable, including a closure, but the grid is using the keys because they are different to the index values. This one might take me a bit longer to fix.

The table is on all of the grid-based graph pages: [http://www.goat1000.com/svggraph-line.php#grid-based-graph-options](http://www.goat1000.com/svggraph-line.php#grid-based-graph-options)

**Comment by goat1000** [comment]
Changeset 92464643e53b739eb61ddf07f17a23c9c910f9ca should fix the callback not working when there are associative/structured data keys.

**Comment by kamilmmach** [comment]
Greetings!
Is there a chance that arbitrary tick labels will be supported sometime in the future?

I'm dealing with line graphs that often have limitation in terms of min/max X values to show and making the grid divide in a way that would show values that I want and also respecting the min/max is impossible in some cases. 

Arbitrary tick labels would likely solve my problem of generating labels with neat values, but maybe there is an option that I overlooked? Like creating a grid in such way that the grid_division would start at position different than axis_min and not grow when it would exceed axis_max.

To put it visually this is what I would want to achieve:
![obraz](https://user-images.githubusercontent.com/74183333/117274077-012f4a00-ae5d-11eb-8615-1ba2db68ccc2.png)

**Comment by goat1000** [comment]
So you would like to be able to position the tick marks explicitly, is that correct? Something like this:
````
$settings['axis_ticks_x'] = [4, 20, 40, 60, 80, 98];
````
With the marks being the correct distance apart for the given values (16 units, 20 units, 20 units, 20 units, 18 units) and overriding the minimum/maximum values.

This doesn't sound too difficult, though the configuration for multiple Y-axes might get a bit complicated.

**Comment by kamilmmach** [comment]
Yes, this is indeed what would work for me! I would always deal with constant distances between tick marks, but those tick marks may not align with axis_min_h and axis_max_h (as in example), hence this solution works!

My use case would be fulfilled, but I think ideally they should not override the minimum/maximum values, but rather indicate only where the ticks/subdivision should be. I think if a tick value is outside of the min/max scope it should be discarded. And the distances between tick marks may also vary - at least these are the standard features of arbitrary axis ticks. But I am aware this may be a lot harder to implement.

And yes, thinking about it now, if Y-axes were to support this feature as well it can probably get a bit messy.

**Comment by kamilmmach** [comment]
This looks amazing and seems to encompass a great part of arbitrary ticks!
Thanks for spending time on this!

**Comment by goat1000** [comment]
Commit 4f2e92d adds support for arbitrary ticks using the options `axis_ticks_x` and `axis_ticks_y`. The value of the option should be an array of values at which to place ticks, or for multiple Y-axes it can be an array of arrays.

Subdivisions are not supported, but date/time keys are:
```
  $options['axis_ticks_x'] = [
    '2020-06-19T00:00:00Z',
    '2020-06-20T13:20:00Z',
    '2020-06-21T12:00:00Z',
    '2020-06-21T16:00:00Z',
    '2020-06-21T20:00:00Z',
  ];
```

The custom tick positions don't change the ends of the axis, for that you should use the `axis_min_h`, `axis_max_h`, `axis_min_v` and `axis_max_v` options. If all of your tick positions are off the graph SVGGraph will throw an exception.

**Comment by kamilmmach** [comment]
I tried it for some of my graphs and it works fantastic! Thanks for adding this feature!
